hypervisor/arch/x86/page.c

```diff
@@ -65,3 +65,132 @@ const struct memory_ops ppt_mem_ops = {
+#define TRUSTY_PGTABLE_PAGE_NUM(size)	\
+(TRUSTY_PML4_PAGE_NUM(size) + TRUSTY_PDPT_PAGE_NUM(size) + TRUSTY_PD_PAGE_NUM(size) + TRUSTY_PT_PAGE_NUM(size))
+
+static struct page uos_sworld_pgtable_pages[CONFIG_MAX_VM_NUM - 1U][TRUSTY_PGTABLE_PAGE_NUM(TRUSTY_RAM_SIZE)];
+
+/* ept: extended page table*/
+static union pgtable_pages_info ept_pages_info[CONFIG_MAX_VM_NUM] = {
+	{
+		.ept = {
+			.top_address_space = EPT_ADDRESS_SPACE(CONFIG_SOS_RAM_SIZE),
+			.nworld_pml4_base = vm0_pml4_pages,
+			.nworld_pdpt_base = vm0_pdpt_pages,
+			.nworld_pd_base = vm0_pd_pages,
+			.nworld_pt_base = vm0_pt_pages,
+		},
+	},
+};
+
+static inline uint64_t ept_get_default_access_right(void)
+{
+	return EPT_RWX;
+}
+
+static inline uint64_t ept_pgentry_present(uint64_t pte)
+{
+	return pte & EPT_RWX;
+}
+
+static inline struct page *ept_get_pml4_page(const union pgtable_pages_info *info, __unused uint64_t gpa)
+{
+	struct page *page;
+	if (gpa < TRUSTY_EPT_REBASE_GPA) {
+		page = info->ept.nworld_pml4_base;
+	} else {
+		page = info->ept.sworld_pgtable_base;
+	}
+	(void)memset(page, 0U, PAGE_SIZE);
+	return page;
+}
+
+static inline struct page *ept_get_pdpt_page(const union pgtable_pages_info *info, uint64_t gpa)
+{
+	struct page *page;
+	if (gpa < TRUSTY_EPT_REBASE_GPA) {
+		page = info->ept.nworld_pdpt_base + (gpa >> PML4E_SHIFT);
+	} else {
+		page = info->ept.sworld_pgtable_base + TRUSTY_PML4_PAGE_NUM(TRUSTY_EPT_REBASE_GPA) +
+			((gpa - TRUSTY_EPT_REBASE_GPA) >> PML4E_SHIFT);
+	}
+	(void)memset(page, 0U, PAGE_SIZE);
+	return page;
+}
+
+static inline struct page *ept_get_pd_page(const union pgtable_pages_info *info, uint64_t gpa)
+{
+	struct page *page;
+	if (gpa < TRUSTY_EPT_REBASE_GPA) {
+		page = info->ept.nworld_pd_base + (gpa >> PDPTE_SHIFT);
+	} else {
+		page = info->ept.sworld_pgtable_base + TRUSTY_PML4_PAGE_NUM(TRUSTY_EPT_REBASE_GPA) +
+			TRUSTY_PDPT_PAGE_NUM(TRUSTY_EPT_REBASE_GPA) + ((gpa - TRUSTY_EPT_REBASE_GPA) >> PDPTE_SHIFT);
+	}
+	(void)memset(page, 0U, PAGE_SIZE);
+	return page;
+}
+
+static inline struct page *ept_get_pt_page(const union pgtable_pages_info *info, uint64_t gpa)
+{
+	struct page *page;
+	if (gpa < TRUSTY_EPT_REBASE_GPA) {
+		page = info->ept.nworld_pt_base + (gpa >> PDE_SHIFT);
+	} else {
+		page = info->ept.sworld_pgtable_base + TRUSTY_PML4_PAGE_NUM(TRUSTY_EPT_REBASE_GPA) +
+			TRUSTY_PDPT_PAGE_NUM(TRUSTY_EPT_REBASE_GPA) + TRUSTY_PD_PAGE_NUM(TRUSTY_EPT_REBASE_GPA) +
+			((gpa - TRUSTY_EPT_REBASE_GPA) >> PDE_SHIFT);
+	}
+	(void)memset(page, 0U, PAGE_SIZE);
+	return page;
+}
+
+static struct memory_ops ept_mem_ops[CONFIG_MAX_VM_NUM] = {
+	{
+		.info = &ept_pages_info[0],
+		.get_default_access_right = ept_get_default_access_right,
+		.pgentry_present = ept_pgentry_present,
+		.get_pml4_page = ept_get_pml4_page,
+		.get_pdpt_page = ept_get_pdpt_page,
+		.get_pd_page = ept_get_pd_page,
+		.get_pt_page = ept_get_pt_page,
+	}
+};
+
+struct memory_ops *init_ept_mem_ops(uint16_t vm_id)
+{
```

Comment:
use vm_id parameter may bring challenge on security and FuSa if vm_id exceed the range.
How about to use  &ept_mem_ops[vm_id].info, and &ept_pages_info[vm_id]   as parameter in caller side.

Reply:
We didn't know the vm_id in the caller side sometimes, ie, in add_pml4e/pdpte/pde.
We would call this function in create_vm after get the valid vm_id. Could we make this as pre-assumption ?